Make TORCH_BOX support const ref arguments

mikaylagawarecki opened the pull request:
This PR removes const and reference for arguments in `unbox_type_t` before the type mapper (`UnboxType`) is applied

For a given argument `a` of type `T` in the function schema, `unbox_type_t` is used to map `T->T1` in order to determine which `to<T1>(a)` to call to convert that StableIValue into a value in the tuple of arguments passed to the function. 

- stripping & from T& to call `to<T>` rather than `to<T&>` is ok because the boxing kernel needs to take ownership of the arguments in the argument tuple
- stripping `const` from `const T` is ok because if the function has the type `const T` for an arg, the StableIValue passed in should be `T` (`to<const T>` is not meaningful)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #169563
* #169385
* #168380

mikaylagawarecki committed:
Make TORCH_BOX support const ref types

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Make TORCH_BOX support const ref types"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Make TORCH_BOX support const ref types"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Make TORCH_BOX support const ref types"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Make TORCH_BOX support const ref types"




[ghstack-poisoned]

albanD reviewed (approved):
Sounds good!
Thanks for the extensive testing on this!

mikaylagawarecki committed:
Update on "Make TORCH_BOX support const ref arguments"

This PR removes const and reference for arguments in `unbox_type_t` before the type mapper (`UnboxType`) is applied

For a given argument `a` of type `T` in the function schema, `unbox_type_t` is used to map `T->T1` in order to determine which `to<T1>(a)` to call to convert that StableIValue into a value in the tuple of arguments passed to the function. 

- stripping & from T& to call `to<T>` rather than `to<T&>` is ok because the boxing kernel needs to take ownership of the arguments in the argument tuple
- stripping `const` from `const T` is ok because if the function has the type `const T` for an arg, the StableIValue passed in should be `T` (`to<const T>` is not meaningful)




[ghstack-poisoned]
